submission/races.go

```diff
@@ -0,0 +1,284 @@
+	var s safeSubmissionState
+	s.logToGroups = ctpolicy.GroupByLogs(groups)
+	s.groupNeeds = make(map[string]int)
+	for _, g := range groups {
+		s.groupNeeds[g.Name] = g.MinInclusions
+	}
+	s.results = make(map[string]*submissionResult)
+	s.cancels = make(map[string]context.CancelFunc)
+	return &s
+}
+
+// request includes empty submissionResult in the set, returns whether the entry wasn't requested (added) before.
+func (sub *safeSubmissionState) request(logURL string, cancel context.CancelFunc) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sub.results[logURL] != nil {
+		// Already requested.
+		return false
+	}
+	sub.results[logURL] = &submissionResult{}
+	isAwaited := false
+	for g := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[g] > 0 {
+			isAwaited = true
+			break
+		}
+	}
+	if !isAwaited {
+		// No groups expecting result from this Log.
+		return false
+	}
+	sub.cancels[logURL] = cancel
+	return true
+}
+
+// returns whether the Log has already been requested.
+func (sub *safeSubmissionState) has(logURL string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	return sub.results[logURL] != nil
+}
+
+// Gets SCT-result. Writes it down if it is error or awaited-SCT. Re-calculates group-completion and cancels any running but not-awited-anymore Log-requests.
+func (sub *safeSubmissionState) setResult(logURL string, sct *ct.SignedCertificateTimestamp, err error) {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sct == nil {
+		sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		return
+	}
+	// If at least one group needs that SCT, result is set. Otherwise dumped.
+	for groupName := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[groupName] > 0 {
+			sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		}
+		sub.groupNeeds[groupName]--
+	}
+
+	// Cancel any pending Log-requests for which there's no more awaiting Log-group.
+	for logURL, groupSet := range sub.logToGroups {
+		isAwaited := false
+		for g := range groupSet {
+			if sub.groupNeeds[g] > 0 {
+				isAwaited = true
+				break
+			}
+		}
+		if !isAwaited && sub.cancels[logURL] != nil {
+			sub.cancels[logURL]()
+			sub.cancels[logURL] = nil
+		}
+	}
+}
+
+// returns whether the group has stopped waiting for more SCTs.
+func (sub *safeSubmissionState) groupComplete(groupName string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	needs, ok := sub.groupNeeds[groupName]
+	if !ok {
+		return true
+	}
+	return needs <= 0
+}
+
+func (sub *safeSubmissionState) collectSCTs() []*AssignedSCT {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	scts := []*AssignedSCT{}
+	for logURL, r := range sub.results {
+		if r != nil && r.sct != nil {
+			scts = append(scts, &AssignedSCT{LogURL: logURL, SCT: r.sct})
+		}
+	}
+	return scts
+}
+
+// postInterval calculates duration for consequent call. For first parallelStart calls duration is 0, while every next one gets additional dur interval.
+func postInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
```

Comment:
Not sure about "post" here. HTTP method? Might be clearer to use submit / submission everywhere.

Reply:
Renamed.